### ISSUE TITLE
Compilation Messages and Annoying Repetitive Toasts

### ISSUE THREAD

**Issue by stevepop** [JeffreyWay/laravel-mix#120]
I am not sure what is going but when I run `npm run dev`. The console keeps displaying compilation and emitting messages. Every time this happens my mac toast notification keeps popping up with 'build successful' even when I am not doing anything. This can get quite annoying and obtrusive.

See image below
<img width="623" alt="continous_compile" src="https://cloud.githubusercontent.com/assets/526063/22102547/bc8dcb76-de2f-11e6-8f36-9090be418e46.png">

**Comment by stevepop** [comment]
Am I the only one experiencing this? No response to this issue yet...

**Comment by kaidesu** [comment]
I get the same thing as well. I've noticed it does this when I use the `copy` method within my mix file.

**Comment by JeffreyWay** [comment]
@stevepop Can you give me more information? What does your webpack.mix.js file look like? I'm thinking that you have something there that's causing Mix to recompile recursively.

**Comment by stevepop** [comment]
@JeffreyWay This is my webpack.mix.js
```
let mix = require('laravel-mix').mix;

/*
 |--------------------------------------------------------------------------
 | Mix Asset Management
 |--------------------------------------------------------------------------
 |
 | Mix provides a clean, fluent API for defining some Webpack build steps
 | for your Laravel application. By default, we are compiling the Sass
 | file for your application, as well as bundling up your JS files.
 |
 */

mix.js('resources/assets/js/frontend.js', 'dist/js');

mix.js('resources/assets/js/backend.js', 'dist/backend/js')
   	.sass('resources/assets/sass/app.scss', 'dist/css')
	.sass('resources/assets/sass/backend.scss', 'dist/backend/');
	.copy('resources/assets/libs/font-awesome/fonts', 'public/dist/css/fonts');

mix.extract(['jquery', 'vue', 'axios', 'lodash', 'moment', 'bootstrap']);
```
Based in what @kaidesu mentioned above, I commented out the line,
`.copy('resources/assets/libs/font-awesome/fonts', 'public/dist/css/fonts');`
but the situation was the same.  Mix just keeps compiling until I cancel the operation.
Thanks for getting back to me on this.

**Comment by JeffreyWay** [comment]
Can you track it down to which mix.* call is causing the repeating compilation? If you reduce it to just mix.js(), does it still repeat?

**Comment by JeffreyWay** [comment]
To debug this, I need someone to give me a simple, reproducible example of how to get these repeated notifications.

**Comment by stevepop** [comment]
@JeffreyWay, I have tracked the issue down as you suggested. This the the webpack.mix.js that works without recursive compilation.
```
let mix = require('laravel-mix').mix;

/*
 |--------------------------------------------------------------------------
 | Mix Asset Management
 |--------------------------------------------------------------------------
 |
 | Mix provides a clean, fluent API for defining some Webpack build steps
 | for your Laravel application. By default, we are compiling the Sass
 | file for your application, as well as bundling up your JS files.
 |
 */

mix.js('resources/assets/js/frontend.js', 'dist/js')
	.js('resources/assets/js/backend.js', 'dist/backend/js')
	//.sass('resources/assets/sass/app.scss', 'dist/css')
	.sass('resources/assets/sass/backend.scss', 'dist/backend/')
 	.copy('resources/assets/libs/font-awesome/fonts', 'public/dist/css/fonts');

mix.sourceMaps();
mix.extract(['jquery', 'vue', 'axios', 'lodash', 'moment', 'bootstrap']);
```

As soon as I comment out the line;
`.sass('resources/assets/sass/app.scss', 'dist/css')`.  It starts again. So I went to my app.scss file and began debugging line by line until I got to this the import for fontawesome.scss;

`@import "vendor/font-awesome/scss/font-awesome.scss";`
Once I commented this line out, everything worked!  I don't know what the issue is but I am glad I am able to reproduce this issue at your suggestion. 
Are you able to reproduce this issue with Font Awesome?  Thanks!

**Comment by JeffreyWay** [comment]
I can't imagine why that's happening. Hmm.. 

Can you confirm that you're on the latest version of Laravel Mix? 0.5.2?

**Comment by stevepop** [comment]
@JeffreyWay, here is an update. I had the following in my font-awesome scss file.
```
/*!
 *  Font Awesome 4.7.0 by @davegandy - http://fontawesome.io - @fontawesome
 *  License - http://fontawesome.io/license (Font: SIL OFL 1.1, CSS: MIT License)
 */

@import "variables";
@import "mixins";
//@import "path";
@import "core";
@import "larger";
@import "fixed-width";
@import "list";
@import "bordered-pulled";
@import "animated";
@import "rotated-flipped";
@import "stacked";
@import "icons";
@import "screen-reader";
```
Everything works when `@import "path" is commented out. Not sure what that means but that is what I have narrowed it down to.
Hmm, I checked my laravel-mix version it seems to be 0.4.0.

**Comment by JeffreyWay** [comment]
Go ahead and bump to 0.5.2 but I don't think that will change this particular issue. 

Hmm - so weird!

**Comment by stevepop** [comment]
Hmm, I updated to Laravel Mix 0.5.2. Now having the following errors
```
✘ steve@steves-devbox  ~/devapps/sites/vue2-naijaprizes   dev-wip ●✚  npm run dev

> @ dev /Users/steve/devapps/sites/vue2-naijaprizes
> cross-env NODE_ENV=development webpack --watch --progress --hide-modules

/Users/steve/devapps/sites/vue2-naijaprizes/webpack.config.js:32
module.exports.context = Mix.paths.root();
                                  ^

TypeError: Cannot read property 'root' of undefined
    at Object.<anonymous> (/Users/steve/devapps/sites/vue2-naijaprizes/webpack.config.js:32:35)
```

**Comment by JeffreyWay** [comment]
Don't forget to update your `webpack.config.js`, if you're using your own copy. 

https://github.com/JeffreyWay/laravel-mix/blob/master/setup/webpack.config.js

**Comment by stevepop** [comment]
Okay, Lavel Mix 0.5.2 installed and working. Issue with font-awesome.scss persists. The following is all that is in font-awesome's path.scss;
```
/* FONT PATH
 * -------------------------- */

@font-face {
  font-family: 'FontAwesome';
  src: url('#{$fa-font-path}/fontawesome-webfont.eot?v=#{$fa-version}');
  src: url('#{$fa-font-path}/fontawesome-webfont.eot?#iefix&v=#{$fa-version}') format('embedded-opentype'),
    url('#{$fa-font-path}/fontawesome-webfont.woff2?v=#{$fa-version}') format('woff2'),
    url('#{$fa-font-path}/fontawesome-webfont.woff?v=#{$fa-version}') format('woff'),
    url('#{$fa-font-path}/fontawesome-webfont.ttf?v=#{$fa-version}') format('truetype'),
    url('#{$fa-font-path}/fontawesome-webfont.svg?v=#{$fa-version}#fontawesomeregular') format('svg');
//  src: url('#{$fa-font-path}/FontAwesome.otf') format('opentype'); // used when developing fonts
  font-weight: normal;
  font-style: normal;
}
```
If I comment everything out, Mix working without the recompile issues but once I uncomment, it starts again.

**Comment by QWp6t** [comment]
`.copy('resources/assets/libs/font-awesome/fonts', 'public/dist/css/fonts');`

^You shouldn't have to do this. 

With a properly set up webpack config, it should copy any assets used within your css. In the case of font-awesome, just do something like this in you scss...

```
$fa-font-path: '~font-awesome/fonts'; // this would normally go in a settings/variables file
@import "~font-awesome/scss/font-awesome"; // this would go with the rest of your imports
```

This works because of the [file-loader](https://webpack.js.org/loaders/file-loader/).
See: https://github.com/JeffreyWay/laravel-mix/blob/master/setup/webpack.config.js#L112-L119

**Comment by kaidesu** [comment]
I can confirm that I get the same issue as reported by @stevepop - it does seem to boil down to the `src` property for fonts causing the watcher to constantly compile.

```
@font-face {
  font-family: 'fusion';

  // src:  url('#{$icomoon-font-path}/fusion.eot?d8lzr');
  // src:  url('#{$icomoon-font-path}/fusion.eot?d8lzr#iefix') format('embedded-opentype'),
  //   url('#{$icomoon-font-path}/fusion.ttf?d8lzr') format('truetype'),
  //   url('#{$icomoon-font-path}/fusion.woff?d8lzr') format('woff'),
  //   url('#{$icomoon-font-path}/fusion.svg?d8lzr#fusion') format('svg');

  font-weight: normal;
  font-style: normal;
}
```

**Comment by stevepop** [comment]
@QWp6t, Thanks for your update. I followed your suggestions and did as follows;
- I removed .copy('resources/assets/libs/font-awesome/fonts', 'public/dist/css/fonts');
- Added $fa-font-path relative to my css ($fa-font-path: '../fonts';)
- Imported form-awesome.css
No more recursive compilation occurred and the fonts were copied over to public/fonts. On my pages however the icons no longer display. I think this was why I ended up having to execute the mix.copy function above. See image below.
![fa_icon](https://cloud.githubusercontent.com/assets/526063/22242140/2ce540f0-e21a-11e6-8f84-7814ac2eb3b9.png)

@JeffreyWay, I think this ticket can now be closed as the original issue reported seems to have been resolved. I guess I will have to think of an alternative way to get the fonts properly referenced.
